src/security/builtin_plugins/authentication/src/authentication.c

```diff
@@ -0,0 +1,3514 @@
+            /* We'll only return the exception of the last one, if it failed. */
+            DDS_Security_Exception_reset(ex);
+            result = verify_certificate(identityCert, trusted_ca_list->buffer[i], ex);
+            if (result == DDS_SECURITY_VALIDATION_OK) {
+                break;
+            }
+        }
+    }
+
+    if (result != DDS_SECURITY_VALIDATION_OK) {
+        goto err_inv_identity_cert;
+    }
+
+    result = check_certificate_expiry( identityCert, ex);
+    if (result != DDS_SECURITY_VALIDATION_OK) {
+        goto err_inv_identity_cert;
+    }
+
+    c_perm = DDS_Security_DataHolder_find_binary_property(token, "c.perm");
+    if (!c_perm) {
+        result = DDS_SECURITY_VALIDATION_FAILED;
+        DDS_Security_Exception_set(ex, DDS_AUTH_PLUGIN_CONTEXT, DDS_SECURITY_ERR_UNDEFINED_CODE, (int)result,
+                "begin_handshake_reply: HandshakeMessageToken property c.perm missing");
+        goto err_no_c_perm;
+    }
+
+    if (c_perm->value._length > 0) {
+        ddsrt_free(relation->remoteIdentity->permissionsDocument);
+        relation->remoteIdentity->permissionsDocument = string_from_data(c_perm->value._buffer, c_perm->value._length);
+    }
+
+    c_pdata = DDS_Security_DataHolder_find_binary_property(token, "c.pdata");
+    if (!c_pdata) {
+        result = DDS_SECURITY_VALIDATION_FAILED;
+        DDS_Security_Exception_set(ex, DDS_AUTH_PLUGIN_CONTEXT, DDS_SECURITY_ERR_UNDEFINED_CODE, (int)result,
+                "begin_handshake_reply: HandshakeMessageToken property c.pdata missing");
+        goto err_no_c_pdata;
+    }
+
+    result = validate_pdata(&c_pdata->value, identityCert, ex);
+    if (result != DDS_SECURITY_VALIDATION_OK) {
+        goto err_inv_pdata;
+    }
+
+    c_dsign_algo = DDS_Security_DataHolder_find_binary_property(token, "c.dsign_algo");
+    if (!c_dsign_algo) {
+        result = DDS_SECURITY_VALIDATION_FAILED;
+        DDS_Security_Exception_set(ex, DDS_AUTH_PLUGIN_CONTEXT, DDS_SECURITY_ERR_UNDEFINED_CODE, (int)result,
+                "begin_handshake_reply: HandshakeMessageToken property c.dsign_algo missing");
+        goto err_no_c_dsign_algo;
+    }
+
+    dsignAlgoKind = get_dsign_algo_from_string((const char *)c_dsign_algo->value._buffer);
+    if (dsignAlgoKind == AUTH_ALGO_KIND_UNKNOWN) {
+        result = DDS_SECURITY_VALIDATION_FAILED;
+        DDS_Security_Exception_set(ex, DDS_AUTH_PLUGIN_CONTEXT, DDS_SECURITY_ERR_UNDEFINED_CODE, (int)result,
+                "begin_handshake_reply: HandshakeMessageToken property c.dsign_algo not supported");
+        goto err_no_c_dsign_algo;
+    }
+
+    c_kagree_algo = DDS_Security_DataHolder_find_binary_property(token, "c.kagree_algo");
+    if (!c_kagree_algo) {
+        result = DDS_SECURITY_VALIDATION_FAILED;
+        DDS_Security_Exception_set(ex, DDS_AUTH_PLUGIN_CONTEXT, DDS_SECURITY_ERR_UNDEFINED_CODE, (int)result,
+                "begin_handshake_reply: HandshakeMessageToke
```

Comment:
* You now add the expiry listener before generating the shared secret, but the error handling labels are reversed;
* If generate_shared_secret fails I presume you need to undo the effect of ``add_validity_end_trigger``.

It does seem to me that the original order was smarter, but that is beside the point, which is that the exception handling should be correct.

Reply:
I thought it would be hard to revert handshake finalization at the beginning.
I've changed the order.